src/Cake.NuGet/Install/Extensions/NuGetPackageManagerExtensions.cs

```diff
@@ -0,0 +1,258 @@
+
+                // If not found, check remote sources.
+                if (sourceRepository == null)
+                {
+                    sourceRepository = await GetSourceRepositoryAsync(packageIdentity, sources, logger, token);
+                }
+
+                // If still not found, we just throw an exception.
+                if (sourceRepository == null)
+                {
+                    throw new InvalidOperationException(
+                        $"Unable to find version '{packageIdentity.Version}' of package '{packageIdentity.Id}'.");
+                }
+
+                return new[] { NuGetProjectAction.CreateInstallProjectAction(packageIdentity, sourceRepository, nuGetProject) };
+            }
+
+            const bool downgradeAllowed = false;
+
+            // Step-1 : Get metadata resources using gatherer
+            var targetFramework =
+                nuGetProject.GetMetadata<NuGetFramework>(NuGetProjectMetadataKeys.TargetFramework);
+            var primaryPackages = new List<PackageIdentity> { packageIdentity };
+            var allSources = new HashSet<SourceRepository>(localSources, new SourceRepositoryComparer());
+            allSources.AddRange(sources);
+            var gatherContext = new GatherContext()
+            {
+                InstalledPackages = Array.Empty<PackageIdentity>(),
+                PrimaryTargets = primaryPackages,
+                TargetFramework = targetFramework,
+                PrimarySources = allSources.ToList(),
+                AllSources = allSources.ToList(),
+                PackagesFolderSource = packageManager.PackagesFolderSourceRepository,
+                ResolutionContext = resolutionContext,
+                AllowDowngrades = downgradeAllowed,
+                ProjectContext = nuGetProjectContext
+            };
+
+            var availablePackageDependencyInfoWithSourceSet = await ResolverGather.GatherAsync(gatherContext, token);
+
+            if (!availablePackageDependencyInfoWithSourceSet.Any())
+            {
+                throw new InvalidOperationException($"Unable to gather dependency information for package '{packageIdentity}'");
+            }
+
+            // Prune the results down to only what we would allow to be installed
+            // Keep only the target package we are trying to install for that Id
+            var prunedAvailablePackages = PrunePackageTree.RemoveAllVersionsForIdExcept(
+                availablePackageDependencyInfoWithSourceSet,
+                packageIdentity).ToList();
+
+            if (!resolutionContext.IncludePrerelease)
+            {
+                prunedAvailablePackages = PrunePackageTree.PrunePreleaseForStableTargets(
+                    prunedAvailablePackages,
+                    new[] { packageIdentity },
+                    new[] { packageIdentity }).ToList();
+            }
+
+            // Step-2 : Call PackageResolver.Resolve to get new list of installed packages
```

Comment:
Could we break out step 2 into it's own method somehow? A lot is happening here and the method would probably benefit from being smaller.